File: Selftest1_compact.m
function Selftest1_compact(fileName)
if nargin<1, fileName='test0929.wav'; end

% ==== 参数 ====
N=512; c=343; Dia=0.08; r=Dia/2; M=8;
rhoL=0.01; rhoH=0.5; Ksmooth=5;
cohTh=1.7; vadTh=6; noiseA=0.98; srpA=0.85;
fmin=300; fmax=3400; nR=81; 
rhoGrid=linspace(rhoL,rhoH,nR); %根据距离上限和距离下限生成均分的网格
dTheta=-6:2:6;

% ==== 文件/阵列 ====
info=audioinfo(fileName); fs=info.SampleRate; T=info.TotalSamples/fs;
afr=dsp.AudioFileReader('Filename',fileName,'SamplesPerFrame',N);
ang=(0:M-1)*45; mic=[r*cosd(ang(:)) r*sind(ang(:))]; Pairs=nchoosek(1:M,2);

% ==== 图 ====
f=figure('Name','GCC角度 + SRP距离','NumberTitle','off'); tl=tiledlayout(f,2,1,'Padding','compact','TileSpacing','compact');
ax1=nexttile(tl,1); hold(ax1,'on'); h1=plot(ax1,nan,nan,'b--'); h2=plot(ax1,nan,nan,'r-'); xlim(ax1,[0 T]); ylim(ax1,[-180 180]); grid(ax1,'on'); title(ax1,'Angle vs Time');
ax2=nexttile(tl,2); hold(ax2,'on'); h3=plot(ax2,nan,nan,'g--'); h4=plot(ax2,nan,nan,'m-'); xlim(ax2,[0 T]); ylim(ax2,[0 1]); grid(ax2,'on'); title(ax2,'Distance vs Time');
axp=polaraxes('Parent',f,'Position',[0.65,0.55,0.30,0.38]); hold(axp,'on'); axp.ThetaZeroLocation='right'; axp.ThetaDir='counterclockwise'; axp.ThetaTick=0:30:330; axp.RLim=[0 r*1.2]; title(axp,'Angle (fixed radius)');
for m=1:M, polarplot(axp,deg2rad(ang(m)),r,'k^','MarkerFaceColor','y','MarkerSize',8); end
hl=polarplot(axp,nan,nan,'r-','LineWidth',1.5); hm=polarplot(axp,nan,nan,'ro','MarkerSize',8,'LineWidth',2);

% ==== 缓存 ====
t=[]; Adeg=[]; Rho=[]; treal=0; prevRho=NaN; tauMax=Dia/c; srpAcc=zeros(1,nR); noiseRMS=[]; win=0.5*(1-cos(2*pi*(0:N-1)'/N));

% ==== 主循环 ====
while ~isDone(afr)
    x=afr(); xw=x.*repmat(win,1,M); treal=treal+N/fs; %加hann窗实现帧头帧尾平滑过渡
    
    frmRMS=median(sqrt(mean(xw.^2,1))); %计算加窗后信号的均方根
    if isempty(noiseRMS), noiseRMS=frmRMS; end %第一帧将frmRMS作为背景噪声基准
    snrDb=20*log10((frmRMS+1e-12)/(noiseRMS+1e-12)); %计算信噪比
    if snrDb<vadTh, noiseRMS=noiseA*noiseRMS+(1-noiseA)*frmRMS; end %指数移动平均法

    [tdoa, wcoh]=gcc_sub(xw, Pairs, N, fs, tauMax); %计算时延
    [udir,adeg]=dir_est(mic, Pairs, c, tdoa); %最小二乘计算角度
    if sum((pairA(mic,Pairs)*udir).* (c*tdoa))<0, udir=-udir; adeg=wrap180(adeg+180); end %若u方向雨实际点积方向相反，旋转180度

    Xi=fft(xw,N); good = (snrDb>=vadTh) && (median(wcoh)>=cohTh);%判断足够的信噪比和相干度
    if ~good
        rho = tern(isnan(prevRho), 0.20, prevRho);%若不靠谱：判断上一个距离是否为nan，若是为默认值0.20m，若否将沿用上一个距离
    else
        bestIdx=1; %最佳距离网格索引，初设为第一个
        bestU=udir; %最佳方向向量，初设为上一步计算的主方向
        bestA=adeg; %最佳角度，初设为上一步的角度估计
        for th=adeg+dTheta %在角度估计+-6°做2°为步进的微调生成
            u=[cosd(th) sind(th)]; %七个候选方向
            sc=scores_pair(Xi,fs,N,c,mic,u,rhoGrid,fmin,fmax); %计算得分
            tmp=srpA*srpAcc+(1-srpA)*sc; %EMA把历史得分与当前得分融合
            [~,idx]=max(tmp);%找出最大值的索引
            if tmp(idx) > srpAcc(bestIdx), bestIdx=idx; bestU=u; bestA=th; end%更新最佳距离索引，对应的方向bestU=u和角度bestA=th
        end
        sc = scores_pair(Xi,fs,N,c,mic,bestU,rhoGrid,fmin,fmax); 
        srpAcc=srpA*srpAcc+(1-srpA)*sc; [~,bestIdx]=max(srpAcc);
        rho=rhoGrid(bestIdx); udir=bestU; adeg=bestA; %使用刚才更新的最佳方向bestU再精算一次，并更新对应距离，角度，方向
    end

    t(end+1)=treal; Adeg(end+1)=adeg; Rho(end+1)=rho; %将当前时间戳，角度，距离进行存储
    A1=movmean(Adeg,Ksmooth); R1=movmean(Rho,Ksmooth); %滑动平均

    set(hl,'ThetaData',[deg2rad(adeg) deg2rad(adeg)],'RData',[0 r]); set(hm,'ThetaData',deg2rad(adeg),'RData',r);
    set(h1,'XData',t,'YData',Adeg); set(h2,'XData',t,'YData',A1);
    set(h3,'XData',t,'YData',Rho);  set(h4,'XData',t,'YData',R1);
    drawnow limitrate; prevRho=rho; pause(N/fs);
end
release(afr); end

% ==== 子函数 ====
function [tdoa,w]=gcc_sub(xw,Pairs,N,fs,tauMax)
P=size(Pairs,1); tdoa=zeros(P,1); w=ones(P,1);
for p=1:P
    i=Pairs(p,1); j=Pairs(p,2);
    Xi=fft(xw(:,i),N); Xj=fft(xw(:,j),N); R=Xi.*conj(Xj); R=R./(abs(R)+eps);
    cc=ifft(R,N,'symmetric'); cc=fftshift(cc);
    [pk,idx]=max(abs(cc)); d=(parab(cc,idx)); lag=(idx-1)-N/2 + d; %对互相关序列做抛物线插值后求时延
    med=median(abs(cc))+eps; w(p)=max(0.1,min(5.0,pk/med)); %找互相关序列的中位数med，将pk/med的结果限制在[0.1,5.0]的范围内，防止极大极小值
    t=lag/fs; tdoa(p)=max(-tauMax,min(t,tauMax));%计算时间差t=lag/fs，t不能超过阵列直径所限制的实验极值[-Dia/c,Dia/c]
end
end

function d=parab(y,i)
N=numel(y); iL=mod(i-2,N)+1; i0=mod(i-1,N)+1; iR=mod(i,N)+1;
d=(y(iL)-y(iR))/(2*(y(iL)-2*y(i0)+y(iR)+eps));
end

function A=pairA(mic,Pairs)
P=size(Pairs,1); A=zeros(P,2);
for p=1:P, A(p,:)=mic(Pairs(p,2),:)-mic(Pairs(p,1),:); end
end

function [u,adeg]=dir_est(mic,Pairs,c,tdoa)
A=pairA(mic,Pairs); b=c*tdoa; uEst=A\b; u=uEst/(norm(uEst)+1e-12); adeg=atan2d(u(2),u(1));%最小二乘法计算声源方位向量u，二维反正切计算角度
end

function sc=scores_pair(Xi,fs,N,c,mic,u,rhos,fmin,fmax)
M=size(mic,1); fk=(0:N-1)'; fk=fk*(fs/N); band=(fk>=fmin & fk<=fmax); fk=fk(band);%获得所有频率点后截取人声区间
X=Xi(band,:); X=X./(abs(X)+eps);%对应频带的频谱并作归一化
Pairs=nchoosek(1:M,2); P=size(Pairs,1); %获取麦克风对矩阵
R=complex(zeros(numel(fk),P)); %互谱矩阵，存放第p对麦在第f个频率点的互谱
for p=1:P, i=Pairs(p,1); 
    j=Pairs(p,2); 
    R(:,p)=X(:,i).*conj(X(:,j)); %逐对计算互谱
end
sc=zeros(1,numel(rhos)); u=reshape(u,1,[]);%sc会保存“假设声源在方向 u 且距离 rhos(k) 处时”的 SRP-PHAT 得分；将列向量u转为行向量
for k=1:numel(rhos)
    s=rhos(k)*u; %假设声源在方向 u 且距离 rhos(k) 处
    D=mic-repmat(s,M,1); %计算每只麦克风相对声源的向量差；mic是M*2的阵列坐标，对应每个麦克风的坐标；repmat(s,M,1) 把 s 重复到 M*2，相减后 D(i,:) 就是第 i 支麦到假设声源的向量
    dist=sqrt(sum(D.^2,2)); %距离：距离向量差的模
    tau=dist/c; %计算时延
    tauij=tau(Pairs(:,2))-tau(Pairs(:,1)); %如果声源在s处，信号应该先后到达1，2麦的时间差
    Ftau=repmat(fk,1,numel(tauij)).*repmat((tauij.'),numel(fk),1);%相位补偿指数部分，频率乘以时延
    W=exp(-1i*2*pi*Ftau); %相位补偿：预测时延下频域多出的部分
    sc(k)=real(sum(sum(R.*W))); %理论相位差与实际相位差相乘，越精确相位就越一致，相乘结果越大
end
end

function y=tern(c,a,b), if c, y=a; else, y=b; end, end
function a=wrap180(a), a=mod(a+180,360)-180; end

File: Selftest1_compact.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Python port of Selftest1_compact.m

Features:
- GCC-PHAT TDoA per mic pair
- Least-squares DoA (2D) and 180° disambiguation
- SRP-PHAT distance grid search around estimated angle ±6° at 2° steps
- VAD by SNR + pairwise coherence gate
- EMA smoothing for noise RMS and SRP accumulator
- Live plots: angle/time, distance/time, and polar angle needle

Dependencies: numpy, scipy, matplotlib, (soundfile optional)
Run: python Selftest1_compact.py [path/to/test0929.wav]
"""

import sys
import time
import math
import numpy as np
import matplotlib.pyplot as plt
from typing import Tuple

# Try to import soundfile (for streaming) and fall back to scipy.io.wavfile
try:
    import soundfile as sf
    HAVE_SF = True
except ImportError:
    HAVE_SF = False
    from scipy.io import wavfile

def hann(N: int) -> np.ndarray:
    n = np.arange(N, dtype=float)
    return 0.5 * (1 - np.cos(2 * np.pi * n / N))

def wrap180(a_deg: float) -> float:
    return (a_deg + 180.0) % 360.0 - 180.0

def tern(cond, a, b):
    return a if cond else b

def parab(y: np.ndarray, i: int) -> float:
    N = y.size
    iL = (i - 1 - 1) % N
    i0 = (i - 1) % N
    iR = (i) % N
    denom = (y[iL] - 2*y[i0] + y[iR] + np.finfo(float).eps)
    return (y[iL] - y[iR]) / (2 * denom)

def pairA(mic: np.ndarray, pairs: np.ndarray) -> np.ndarray:
    P = pairs.shape[0]
    A = np.zeros((P, 2), dtype=float)
    for p in range(P):
        i, j = pairs[p]
        A[p, :] = mic[j, :] - mic[i, :]
    return A

def dir_est(mic: np.ndarray, pairs: np.ndarray, c: float, tdoa: np.ndarray) -> Tuple[np.ndarray, float]:
    A = pairA(mic, pairs)
    b = c * tdoa.reshape(-1, 1)
    u_est, *_ = np.linalg.lstsq(A, b, rcond=None)
    u = u_est.flatten()
    u /= (np.linalg.norm(u) + 1e-12)
    adeg = math.degrees(math.atan2(u[1], u[0]))
    return u, adeg

def gcc_sub(xw: np.ndarray, pairs: np.ndarray, N: int, fs: float, tauMax: float) -> Tuple[np.ndarray, np.ndarray]:
    P = pairs.shape[0]
    tdoa = np.zeros(P, dtype=float)
    w = np.ones(P, dtype=float)
    for p in range(P):
        i, j = pairs[p]
        Xi = np.fft.fft(xw[:, i], n=N)
        Xj = np.fft.fft(xw[:, j], n=N)
        R = Xi * np.conj(Xj)
        R /= (np.abs(R) + np.finfo(float).eps)
        cc = np.fft.ifft(R, n=N).real
        cc = np.fft.fftshift(cc)
        idx = np.argmax(np.abs(cc))
        d = parab(cc, idx)
        lag = (idx - (N//2)) + d
        pk = np.max(np.abs(cc))
        med = np.median(np.abs(cc)) + np.finfo(float).eps
        w[p] = max(0.1, min(5.0, pk/med))
        tdoa[p] = max(-tauMax, min(lag/fs, tauMax))
    return tdoa, w

def scores_pair(Xi: np.ndarray, fs: float, N: int, c: float, mic: np.ndarray,
                u: np.ndarray, rhos: np.ndarray, fmin: float, fmax: float) -> np.ndarray:
    M = mic.shape[0]
    fk = np.arange(N, dtype=float)*(fs/N)
    band = (fk>=fmin)&(fk<=fmax)
    fk = fk[band].reshape(-1,1)
    X = Xi[band,:]/(np.abs(Xi[band,:])+np.finfo(float).eps)
    from itertools import combinations
    pairs = np.array(list(combinations(range(M),2)),dtype=int)
    P = pairs.shape[0]
    R = np.empty((fk.shape[0],P),dtype=complex)
    for p,(i,j) in enumerate(pairs):
        R[:,p] = X[:,i]*np.conj(X[:,j])
    scores = np.zeros(rhos.size,dtype=float)
    u_row = u.reshape(1,-1)
    for k,rho in enumerate(rhos):
        s = rho*u_row
        D = mic - s
        dist = np.linalg.norm(D,axis=1)
        tau = dist/c
        tauij = tau[pairs[:,1]] - tau[pairs[:,0]]
        Ftau = fk @ tauij.reshape(1,-1)
        W = np.exp(-1j*2*math.pi*Ftau)
        scores[k] = np.real(np.sum(R*W))
    return scores

def moving_average(x: np.ndarray, k: int) -> np.ndarray:
    if k<=1:
        return x.copy()
    return np.convolve(x, np.ones(k)/k, mode='same')

def Selftest1_compact_py(file_path: str='test0929.wav'):
    # === parameters ===
    N=512; c=343.0; Dia=0.08; r=Dia/2; M=8
    rhoL=0.01; rhoH=0.5; Ksmooth=5
    cohTh=1.7; vadTh=6.0; noiseA=0.98; srpA=0.85
    fmin=300.0; fmax=3400.0; nR=81
    rhoGrid=np.linspace(rhoL,rhoH,nR)
    dTheta=np.arange(-6,8,2)
    # file read setup
    if HAVE_SF:
        info=sf.info(file_path)
        fs=float(info.samplerate); total=info.frames; T=total/fs
        sf_reader=sf.SoundFile(file_path,'r')
    else:
        fs_i,data=wavfile.read(file_path)
        fs=float(fs_i)
        if data.ndim==1:
            data=data[:,None]
        if np.issubdtype(data.dtype,np.integer):
            data=data.astype(np.float32)/np.iinfo(data.dtype).max
        total=data.shape[0]; T=total/fs
        sf_reader=None
    ang=np.arange(M)*45.0
    mic=np.column_stack((r*np.cos(np.deg2rad(ang)),r*np.sin(np.deg2rad(ang))))
    from itertools import combinations
    Pairs=np.array(list(combinations(range(M),2)),dtype=int)
    # plotting
    plt.close('all')
    fig=plt.figure(num='GCC angle + SRP distance',figsize=(10,7))
    gs=fig.add_gridspec(2,2,width_ratios=[3,2],height_ratios=[1,1],wspace=0.25,hspace=0.25)
    ax1=fig.add_subplot(gs[0,0]); ax1.set_title('Angle vs Time'); ax1.set_xlim(0,T); ax1.set_ylim(-180,180); ax1.grid()
    ax2=fig.add_subplot(gs[1,0]); ax2.set_title('Distance vs Time'); ax2.set_xlim(0,T); ax2.set_ylim(0,1); ax2.grid()
    axp=fig.add_subplot(gs[:,1],projection='polar'); axp.set_title('Angle (fixed radius)')
    axp.set_theta_zero_location('E'); axp.set_theta_direction(1); axp.set_thetagrids(np.arange(0,360,30)); axp.set_rlim(0,r*1.2)
    h1,=ax1.plot([],[],linestyle='--'); h2,=ax1.plot([],[],linestyle='-')
    h3,=ax2.plot([],[],linestyle='--'); h4,=ax2.plot([],[],linestyle='-')
    for m in range(M):
        axp.plot([math.radians(ang[m])],[r],marker='^')
    hl,=axp.plot([],[],'-'); hm,=axp.plot([],[],'o')
    # buffers
    t_list=[]; Adeg=[]; Rho=[]; treal=0.0; prevRho=np.nan; tauMax=Dia/c
    srpAcc=np.zeros(nR); noiseRMS=None; win= hann(N).reshape(-1,1)
    
    # frame reader
    def read_frame():
        if HAVE_SF:
            frame = sf_reader.read(N, dtype='float64', always_2d=True)
            if frame.size == 0: return None
        else:
            if read_frame.idx >= total: return None
            i0 = read_frame.idx
            i1 = min(i0 + N, total)
            frame = data[i0:i1]
            read_frame.idx = i1
            if frame.shape[0] < N:
                pad = np.zeros((N - frame.shape[0], frame.shape[1]))
                frame = np.vstack((frame, pad))
    # 共同处理逻辑
        ch = frame.shape[1]
        if ch < M:
            frame = np.tile(frame, (1, int(np.ceil(M / ch))))[:, :M]
        elif ch > M:
            frame = frame[:, :M]
        return frame

# 初始化 idx（必须在函数定义后、循环前添加此行）
    if not HAVE_SF: read_frame.idx = 0
    # main loop
    while True:
        x=read_frame()
        if x is None: break
        xw=x*win; treal+=N/fs
        frmRMS=np.median(np.sqrt(np.mean(xw**2,axis=0)))
        if noiseRMS is None: noiseRMS=frmRMS
        snrDb=20*math.log10((frmRMS+1e-12)/(noiseRMS+1e-12))
        if snrDb<vadTh: noiseRMS=noiseA*noiseRMS+(1-noiseA)*frmRMS
        tdoa,w= gcc_sub(xw,Pairs,N,fs,tauMax)
        udir,adeg= dir_est(mic,Pairs,c,tdoa)
        if np.sum((pairA(mic,Pairs)@udir)*(c*tdoa))<0:
            udir=-udir; adeg=wrap180(adeg+180)
        Xi=np.fft.fft(xw,n=N,axis=0)
        good=(snrDb>=vadTh) and (np.median(w)>=cohTh)
        if not good:
            rho=0.20 if np.isnan(prevRho) else prevRho
        else:
            bestIdx=0; bestU=udir.copy(); bestA=adeg
            for th in adeg + dTheta:
                u=[math.cos(math.radians(th)),math.sin(math.radians(th))]
                sc=scores_pair(Xi,fs,N,c,mic,u,rhoGrid,fmin,fmax)
                tmp=srpA*srpAcc+(1-srpA)*sc
                idx=int(np.argmax(tmp))
                if tmp[idx]>srpAcc[bestIdx]:
                    bestIdx=idx; bestU=u; bestA=th
            sc=scores_pair(Xi,fs,N,c,mic,bestU,rhoGrid,fmin,fmax)
            srpAcc=srpA*srpAcc+(1-srpA)*sc
            bestIdx=int(np.argmax(srpAcc)); rho=float(rhoGrid[bestIdx])
            udir=bestU; adeg=bestA
        t_list.append(treal); Adeg.append(adeg); Rho.append(rho)
        A1=moving_average(np.array(Adeg),Ksmooth)
        R1=moving_average(np.array(Rho),Ksmooth)
        # update plots
        hl.set_data([math.radians(adeg),math.radians(adeg)],[0,r])
        hm.set_data([math.radians(adeg)],[r])
        h1.set_data(t_list,Adeg); h2.set_data(t_list,A1)
        h3.set_data(t_list,Rho); h4.set_data(t_list,R1)
        ax1.set_xlim(0,max(T,t_list[-1])); ax2.set_xlim(0,max(T,t_list[-1]))
        plt.pause(N/fs*0.2)
        prevRho=rho
    if HAVE_SF: sf_reader.close()
    plt.show()

if __name__=="__main__":
    path=sys.argv[1] if len(sys.argv)>1 else "test0929.wav"
    Selftest1_compact_py(path)
